### DISCUSSION TITLE
Can't get addressability to Django

### DISCUSSION THREAD

**Discussion by eddhyne** [microsoft/pylance-release#4415] [Q&A] [answered]
All my references to django imports have sqiggly lines. Error msg is "Import "django.utilis.timezone" could not be resolved.
My pip.list command shows Django 4.2.1 installed in my virtual environment.
I have no Django intellisense
Attached is a screenshot.
![Snag_1c6ccb](https://github.com/microsoft/vscode-python/assets/92999884/45465cc3-ae8f-4c19-a1cc-5f032cf53040)

**Comment by brettcannon** [comment]
I've transferred this to Pylance, but one key detail you cropped out of your screenshot is what environment you have selected (to verify it matches the one you have active in your terminal)?

**Comment by eddhyne** [comment]
Thanks for getting back to me.

How do I verify the one I should be using? The one I am using is the virtual environment I installed packages in. When I add code, where should I be?

***@***.***

Ed Dhyne
Direct: 954-608-6603
***@***.******@***.***>
www.linkedin.com

**Comment by rchiodo** [comment]
We'd need to know the environment selected in VS Code. It's on the bottom right of the window:

![image](https://github.com/microsoft/pylance-release/assets/19672699/6dd47aa3-a2b7-424f-b0ca-81d14b6c53f3)

That's usually the reason why Pylance can't find your imports. It's not using the same environment that you're running with.

But it could be other reasons too.

**Comment by eddhyne** [comment]
Yes that is the virtual environment I install Django in.

***@***.***

Ed Dhyne
Direct: 954-608-6603
***@***.******@***.***>
www.linkedin.com

**Comment by rchiodo** [comment]
Okay the next thing we need is a log of Pylance's analysis. Directions for capturing that log can be found here:
https://github.com/microsoft/pylance-release/blob/main/TROUBLESHOOTING.md#filing-an-issue

**Comment by eddhyne** [comment]
Here is the log and requirements.txt



Ed Dhyne
Direct: 954-608-6603
***@***.******@***.***>
www.linkedin.com

**Reply by rchiodo** [reply]
It looks like you're using e-mail to reply. It doesn't include links in your e-mail in the actual discussion so we can't see the logs.

**Reply by eddhyne** [reply]
I tried dragging the log into this issue. I will try again.
[python_log .docx](https://github.com/microsoft/pylance-release/files/11558361/python_log.docx)
[python_log .docx](https://github.com/microsoft/pylance-release/files/11558383/python_log.docx)

**Reply by rchiodo** [reply]
Can you turn on the trace log setting and try again?

![image](https://github.com/microsoft/pylance-release/assets/19672699/6bbd4c01-f46f-4585-b72b-d1964058dbe0)

And if you could copy the contents into a txt file, that would be much more manageable than a screen shot.

**Reply by eddhyne** [reply]
I copied the trace into a text document. I dragged the icon of the .txtto the opening below and that's what I am sending.
[2023-05-24 151337.129 [info] [Info .txt](https://github.com/microsoft/pylance-release/files/11558541/2023-05-24.151337.129.info.Info.txt)

**Reply by rchiodo** [reply]
According to the log it says this is the python path that Pylance is getting:

```
2023-05-24 15:13:37.211 [info] [Info  - 3:13:37 PM] (33116) Setting pythonPath for service "Edshello_django": "D:\Python311\python.exe"
```

Which doesn't match the one you were using in the original screen shot. 

What does the bottom right of the screen show as the Python environment?

**Reply by eddhyne** [reply]
So how did it get changed from Z:\Edshello_django\.venv?  How do I change it back to this. Thats where I installed Django.

**Reply by rchiodo** [reply]
The bottom of VS Code has a status bar. You can click on the status bar to select the environment that VS code is using. Like so:

![image](https://github.com/microsoft/pylance-release/assets/19672699/a3a118e0-bd8a-44f9-aa3e-a1c2eeae263e)

That lets you pick the environment used inside of VS code. It doesn't automatically pick up the one you're using in your terminal (even if that terminal is running inside of VS code).

**Reply by eddhyne** [reply]
It shows Python 3.11.3('.venv':venv)
This is my 1st Django app I was following Django Tutorial in Visual Studio Code https://code.visualstudio.com/docs/python/tutorial-django. I created my own virtual environment file
and install django into it. Following the tutorial, where should my code go into? 
(.venv) PS Z:\Edshello_django> or PS Z:\Edshello_django> ? How do I fix it?

**Reply by rchiodo** [reply]
You should be able to fix it by picking the virtual environment that you installed Django into. I believe it should have a path like `Z:\Edshello_django\.venv\scripts\python.exe` in the example you gave. 

You'd pick it with the 'Select Interpreter' command or by clicking on the status bar (which basically runs that command).

**Reply by eddhyne** [reply]
However If I pick the one I created, I get all the squiggly error line. How do I fix that?

**Reply by rchiodo** [reply]
The log for the 'Python Language Server' should show a line saying it's using that environment. Something like so:

```
2023-05-24 15:13:37.211 [info] [Info  - 3:13:37 PM] (33116) Setting pythonPath for service "Edshello_django": "Z:\Edshello_django\.venv\scripts\python.exe"
```

If it doesn't say that, then you didn't pick the right environment. 

If it does say that, then you may be hitting a problem where UNC paths don't work. This VS code issue here:
https://github.com/microsoft/vscode/issues/182070

There's an allow list now for UNC paths.
https://code.visualstudio.com/docs/setup/windows#_working-with-unc-paths

**Reply by eddhyne** [reply]
When I hit Ctrl+Shift+p I get a dropdown showing
Python 3.11.3('venv':).\.venv\Scripts\python.exe     Recommended   What is it recommended for?
. When I enter code, I get squiggly line on all Djano imports and no Django intelisense
    Is this .venv only for installing packages in? This has Django 4.2.1 the llatest 1 I installed.
. Python 3.11.3 64-bit D:\Python311\python.exe       Global
    When I select this, I get no squiggly lines and I do get Django intelisense
    I this what I should select to do all my coding, runserver testing, debugging etc?
    This also has This has Django 4.2.1

**Reply by rchiodo** [reply]
You can pick whichever one you want to run/install packages into. We generally recommend using a venv for doing this though. 

I'm not sure what `Z:` is mapped to, maybe it's not a UNC path, but that might be why that venv doesn't work. VS code recently made it harder to use UNC paths because they're considered a security risk. 

You could :
- Use the global python
- Create another venv somewhere on a local disk

**Reply by eddhyne** [reply]
I will use the global python
My z: drive is a mapped drive to a SYNOLOGY NAS. Could that be causing the problem?
I will move the project to a regular Windows drive and delete and recreate the venv to see if it makes a difference. Thanks.

**Reply by eddhyne** [reply]
How can I tell when Visual Studio Code was updated? I think it was yesterday on my PC. It fixed my problem. Its working OK on my NAS Z: drive. See my attachment.
![Snag_3bf529b](https://github.com/microsoft/pylance-release/assets/92999884/a6329040-7c9f-4cec-aef9-c88f822a6410)

**Comment by eddhyne** [comment]
I added all the info below to git but I don’t see my lates update there.

How can I tell when Visual Studio Code update 1.78.2 was installed on my PC. It seams to have fixed my problem. I am working in the project in my NAS Z: drive and everything seam to be working when I am in my .VENV or Global. I will check it out further.

***@***.***


Ed Dhyne
Direct: 954-608-6603
***@***.******@***.***>
www.linkedin.com

**Comment by eddhyne** [comment]
A few years ago I create www.imaginativesystems.com<http://www.imaginativesystems.com> in ASP.NET C#. I would like to replace it with a Python Django version. So far the tutorials I have tried only get me started. They don’t go into horizontal menus and vertical links like on my ISI site. Can you recommend any tutorials, courses, books or manuals that will help me. I would appreciate anything you can offer now that I can develop django projects in Visual Studio Code.

Thanks for everything.

Ed Dhyne
Direct: 954-608-6603
***@***.******@***.***>
www.linkedin.com

**Reply by eddhyne** [reply, marked as answer]
After release update 1.78.2 was installed the other day, my problems have cleared up.
